File: 2015/zig/day1/day1.zig
const std =  @import("std");
const data = @embedFile("../../inputs/input.txt");

pub fn part1() !void {
  var lines = std.mem.tokenize(u8, data, "\n");

  var up: i32 = 0;
  var down: i32 = 0;
  while(lines.next()) |line| {
    // var splits = std.mem.split(u8, line, "");
    for (line) |character| {
      if (character == ')') {
        up += 1;
      } else {
        down += 1;
      }
    }
  }
  std.debug.print("{d}\n", .{down - up});
}

pub fn part2() !void {
  var lines = std.mem.tokenize(u8, data, "\n");

  var pos: i32 = 0;
  var i: i32 = 0;
  while(lines.next()) |line| {
    // var splits = std.mem.split(u8, line, "");
    for (line) |character| {
      i += 1;
      if (character == '(') {
        pos += 1;
      } else {
        pos -= 1;
      }
      if (pos == -1) {
        std.debug.print("{d}\n", .{i});
        return;
      }
    }
  }
}

pub fn main() !void {
  // try part1();
  try part2();
}

File: 2015/go/day1/day1.go
package main

import (
	"errors"
	"fmt"
	"os"
)

func read_file(file_name string) string {
  file, err := os.ReadFile(file_name)
  if err != nil {
    panic(err)
  }
  return string(file)
}

func part1() int {
  data := read_file("../../inputs/input.txt")
  up := 0
  down := 0
  for _, element := range data {
    if element == ')' {
      up += 1
    } else if element == '('{
      down += 1
    }

  }
  return down - up
}

func part2() (int, error) {
  
  data := read_file("../../inputs/input.txt")
  position := 0
  for index, element := range data {
    if element == ')' {
      position -= 1
    } else if element == '('{
      position += 1
    }

    if position == -1 {
      return index + 1, nil
      
    }

  }
  return 0, errors.New("Nope")
}

func main()  {
  // fmt.Println(part1())
  fmt.Println(part2())
}
